parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/ParquetRewriter.java

```diff
@@ -106,48 +109,63 @@ public class ParquetRewriter implements Closeable {
   private int numBlocksRewritten = 0;
   // Reader and relevant states of the in-processing input file
   private final Queue<TransParquetFileReader> inputFiles = new LinkedList<>();
-  // Schema of input files (should be the same) and to write to the output file
-  private MessageType schema = null;
-  private final Map<ColumnPath, ColumnDescriptor> descriptorsMap;
-  // The reader for the current input file
-  private TransParquetFileReader reader = null;
-  // The metadata of current reader being processed
-  private ParquetMetadata meta = null;
-  // created_by information of current reader being processed
-  private String originalCreatedBy = "";
-  // Unique created_by information from all input files
-  private final Set<String> allOriginalCreatedBys = new HashSet<>();
+  private final Queue<TransParquetFileReader> inputFilesToJoin = new LinkedList<>();
+  private MessageType outSchema;
   // The index cache strategy
   private final IndexCache.CacheStrategy indexCacheStrategy;
+  private final boolean overwriteInputWithJoinColumns;
 
   public ParquetRewriter(RewriteOptions options) throws IOException {
+    this.newCodecName = options.getNewCodecName();
+    this.indexCacheStrategy = options.getIndexCacheStrategy();
+    this.overwriteInputWithJoinColumns = options.getOverwriteInputWithJoinColumns();
     ParquetConfiguration conf = options.getParquetConfiguration();
     OutputFile out = options.getParquetOutputFile();
-    openInputFiles(options.getParquetInputFiles(), conf);
-    LOG.info("Start rewriting {} input file(s) {} to {}", inputFiles.size(), options.getParquetInputFiles(), out);
-
-    // Init reader of the first input file
-    initNextReader();
-
-    newCodecName = options.getNewCodecName();
-    List<String> pruneColumns = options.getPruneColumns();
-    // Prune columns if specified
-    if (pruneColumns != null && !pruneColumns.isEmpty()) {
-      List<String> paths = new ArrayList<>();
-      getPaths(schema, paths, null);
-      for (String col : pruneColumns) {
-        if (!paths.contains(col)) {
-          LOG.warn("Input column name {} doesn't show up in the schema of file {}", col, reader.getFile());
-        }
+    inputFiles.addAll(getFileReaders(options.getParquetInputFiles(), conf));
+    inputFilesToJoin.addAll(getFileReaders(options.getParquetInputFilesToJoin(), conf));
+    ensureSameSchema(inputFiles);
+    ensureSameSchema(inputFilesToJoin);
```

Comment:
The `inputFilesToJoin` are the right-side files, right? Why do they need to have the same schema?

Reply:
> The inputFilesToJoin are the right-side files, right? 

Yes

> Why do they need to have the same schema?

If they don't have the same schema, then how do we figure out which columns need to be added? I guess we can have a list of columns from the right provided as a parameter form a user or create a union of all present columns but then we again getting into complicated implementation imo.

Reply:
I thought this works like: file_1<column a, column b> join file_2<column c>, file_3<column d>, then produce file_joined<column a, column b, column c, column d>

Reply:
Sorry, I didn't get your example. What is the schema for file_2 and file_3? Are file_2 and file_3 both in  `inputFilesToJoin`?


Current logic is:
- `inputFiles` is required (left side), a schema for these file must be the same for all them as it was before this PR
- `inputFilesToJoin` is optional (right side), a schema for these file must be the same for all them
- a final merged schema is produced in a constructor
- in case of an overlapping column that present on the left and on the right its values are extracted from the left or from the right depending on `overwriteInputWithJoinColumns` parameter

Reply:
Sorry,  I have updated the above example. So for the above example, we need to do with two steps: 
1. file_1 join file_2 and get file_joined_1
2. file_joined_ 1 join file_3 and get the final result

Reply:
okay, I got it
This functionality is not supported by this PR. It was possible in the original version of joiner but In this PR I wanted to start with the simplest implementation possible. Maybe I can add it later.